cpp/src/replace/replace.cu

```diff
@@ -182,51 +198,139 @@ void replace_nulls_with_column(gdf_size_type size, Type* out_data, gdf_valid_typ
+ */
+/* ----------------------------------------------------------------------------*/
+struct replace_nulls_scalar_kernel_forwarder {
+  template <typename col_type>
+  void operator()(gdf_size_type    nrows,
+                  void*            d_in_data,
+                  gdf_valid_type*  d_in_valid,
+                  const void*      replacement,
+                  void*            d_out_data)
+  {
+    cudf::util::cuda::grid_config_1d grid{nrows, BLOCK_SIZE};
+
+    auto t_replacement = static_cast<const col_type*>(replacement);
+    void *d_replacement = NULL;
+    RMM_TRY(RMM_ALLOC(&d_replacement, sizeof(col_type), 0));
+    CUDA_TRY(cudaMemcpy(d_replacement, t_replacement, sizeof(col_type),
+                        cudaMemcpyHostToDevice));
+
+    replace_nulls_with_scalar<<<grid.num_blocks, BLOCK_SIZE>>>(nrows,
+                                          static_cast<const col_type*>(d_in_data),
+                                          reinterpret_cast<bit_mask_t*>(d_in_valid),
+                                          static_cast<const col_type*>(d_replacement),
+                                          static_cast<col_type*>(d_out_data)
+                                          );
+    RMM_TRY(RMM_FREE(d_replacement, 0));
   }
 };
 
+
+
 } //end anonymous namespace
 
-gdf_error gdf_replace_nulls(gdf_column* col_out, const gdf_column* col_in)
+
+namespace cudf {
+
+gdf_column replace_nulls(const gdf_column& input,
+                         const gdf_column& replacement)
 {
-  GDF_REQUIRE(col_out->dtype == col_in->dtype, GDF_DTYPE_MISMATCH);
-  GDF_REQUIRE(col_in->size == 1 || col_in->size == col_out->size, GDF_COLUMN_SIZE_MISMATCH);
-   
-  GDF_REQUIRE(nullptr != col_in->data, GDF_DATASET_EMPTY);
-  GDF_REQUIRE(nullptr != col_out->data, GDF_DATASET_EMPTY);
-  GDF_REQUIRE(nullptr == col_in->valid || 0 == col_in->null_count, GDF_VALIDITY_UNSUPPORTED);
-
-  cudf::type_dispatcher(col_out->dtype, replace_nulls_kernel_forwarder{},
-                        col_out->size,
-                        col_in->size,
-                        col_out->data,
-                        col_out->valid,
-                        col_in->data);
-  return GDF_SUCCESS;
+  if (input.size == 0) {
+    return cudf::empty_like(input);
+  }
+
+  CUDF_EXPECTS(nullptr != input.data, "Null input data");
+
+  if (input.null_count == 0 || input.valid == nullptr) {
+    return cudf::copy(input);
+  }
+
+  CUDF_EXPECTS(input.dtype == replacement.dtype, "Data type mismatch");
+  CUDF_EXPECTS(replacement.size == 1 || replacement.size == input.size, "Column size mismatch");
+  CUDF_EXPECTS(nullptr != replacement.data, "Null replacement data");
+  CUDF_EXPECTS(nullptr == replacement.valid || 0 == replacement.null_count,
+               "Invalid replacement data");
+
+  gdf_column output = cudf::empty_like(input);
+  output.size = input.size;
+  if (input.size > 0) {
+    const auto byte_width = (input.dtype == GDF_STRING)
```

Comment:
Hmm, maybe we should add this logic to `cudf::allocate_like()` and then replace this whole allocation with a call to it?